### ISSUE TITLE
Intertopic validation for nmf

### ISSUE THREAD

**Issue by Jhumpa** [bmabey/pyLDAvis#122]
Hi,
Is there any python code to validate the NMF topics like intertopic distance.
Please share the code.

**Comment by abeerunscore96** [comment]
I'd be interested in this as well. The only issue I see with the current code is that it requires the weight of each topic to a document to be a probability, i.e. the sum of weights of each topic to a document to sum to 1. LDA ensures that. NMF doesn't. NMF simply weights the topics with no requirement for it to sum to 1. This is easily remedied by standardizing the NMF document embedding i.e. 

nmf_embedding = nmf.transform(tfidf)
nmf_embedding = (nmf_embedding - nmf_embedding.mean(axis=0))/nmf_embedding.std(axis=0)

If someone could point me to where the transform step happens in the pyLDAvis code, I modify that area with the standardization.

**Comment by sahilarora93** [comment]
Hi all
Should it not be the following : 

def _row_norm(dists):
    # row normalization function required
    # for doc_topic_dists and topic_term_dists
    return dists / dists.sum(axis=1)[:, None]

**Comment by abeer-khan** [comment]
@sahilarora93 You're right. We should divide each cell with the sum to make it sum to 1.

**Comment by sahilarora93** [comment]
Also I did it directly and it seems to work. The same normalisation is already part of the pyLDAvis library.

from sklearn.decomposition import NMF
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
import pyLDAvis
import pyLDAvis.sklearn
pyLDAvis.enable_notebook()

tfidf_vectorizer = TfidfVectorizer(stop_words=stopwords, lowercase=True, strip_accents="unicode", tokenizer=custom_tokenizer, 
                            use_idf=True, norm='l2', min_df = 10, ngram_range = (1,1))
dtm_tfidf = tfidf_vectorizer.fit_transform(docs)
print(dtm_tfidf.shape)

# for TFIDF DTM
nmf_tfidf = NMF(init='random', n_components=5, max_iter=100, random_state = 0)
nmf_tfidf.fit(dtm_tfidf)
vis_nmf = pyLDAvis.sklearn.prepare(nmf_tfidf, dtm_tfidf, tfidf_vectorizer)
vis_nmf

Let me know if there is an issue in the same.

**Comment by sahilarora93** [comment]
@Jhumpa @abeer-khan let me know if this works.

**Comment by spencermehta** [comment]
```
H = nmf.components_
H = H / H.sum
nmf.components_ = H
```
This is how I managed to do it.
